### ISSUE TITLE
network splits and 2 miners are unaware of each other - what happens?

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#264]
@ignopeverell from testing #244 it is possible to get a couple of miners into a state where neither know about each other.
Artificial example - they are both seeded off each other but we run them alternately with the other stopped.
The end result is each mining node has a chainstate that is the consistent up to and including `height n` but they both (independently) mine `block n+1` onward.

What is in place to resolve this when the network fixes itself and the nodes can communicate again?

The sync process results in orphan blocks (node with lower difficulty syncing headers from node with higher difficulty).

I'm following through the `locate_headers` code path.
I see the node requesting what appears to be the correct headers based on the locators.
I see it receiving the headers as requested.
But it does not appear to do the right thing with these to "un-fork" at the appropriate height.

**Comment by antiochp** [comment]
Some more context. We see this in the logs of the node that is attempting to sync (peer difficulty is > self difficulty) on startup - 

```
Nov 14 12:53:34.725 DEBG Sync: peer 4974 vs us 4309
Nov 14 12:53:34.725 DEBG Sync: blocks to download 0, block downloading 0
Nov 14 12:53:34.725 DEBG Loc heights: [81, 79, 77, 73, 65, 49, 17]
Nov 14 12:53:34.729 DEBG Asking peer 127.0.0.1:13414 for more block headers starting from 985ce32d at 81.
Nov 14 12:53:34.732 INFO Received 17 block headers
Nov 14 12:53:34.732 INFO Starting validation pipeline for block header 196c174c at 74.
Nov 14 12:53:34.733 INFO Received unfit block header 196c174c at 74: already in store.
Nov 14 12:53:34.733 INFO Starting validation pipeline for block header 3cb4dd99 at 75.
Nov 14 12:53:34.734 INFO Received unfit block header 3cb4dd99 at 75: already in store.
Nov 14 12:53:34.734 INFO Starting validation pipeline for block header a46ea7fb at 76.
Nov 14 12:53:34.735 INFO Received unfit block header a46ea7fb at 76: already in store.
Nov 14 12:53:34.735 INFO Starting validation pipeline for block header b955b4dc at 77.
Nov 14 12:53:34.736 INFO Received unfit block header b955b4dc at 77: already in store.
Nov 14 12:53:34.736 INFO Starting validation pipeline for block header ad096e49 at 78.
Nov 14 12:53:34.737 INFO Received unfit block header ad096e49 at 78: already in store.
Nov 14 12:53:34.737 INFO Starting validation pipeline for block header 727e52c7 at 79.
Nov 14 12:53:34.738 INFO Received unfit block header 727e52c7 at 79: already in store.
Nov 14 12:53:34.738 INFO Starting validation pipeline for block header d7c7848d at 80.
Nov 14 12:53:34.739 INFO Received unfit block header d7c7848d at 80: already in store.
Nov 14 12:53:34.739 INFO Starting validation pipeline for block header 097dee04 at 81.
Nov 14 12:53:34.740 INFO Received unfit block header 097dee04 at 81: already in store.
Nov 14 12:53:34.740 INFO Starting validation pipeline for block header ae04dd78 at 82.
Nov 14 12:53:34.741 INFO Received unfit block header ae04dd78 at 82: already in store.
Nov 14 12:53:34.741 INFO Starting validation pipeline for block header e36fb1b5 at 83.
Nov 14 12:53:34.741 INFO Invalid block header e36fb1b5: Orphan.
Nov 14 12:53:34.741 INFO Starting validation pipeline for block header c1ca07bd at 84.
Nov 14 12:53:34.742 INFO Invalid block header c1ca07bd: Orphan.
Nov 14 12:53:34.742 INFO Starting validation pipeline for block header 27022fb7 at 85.
Nov 14 12:53:34.742 INFO Invalid block header 27022fb7: Orphan.
Nov 14 12:53:34.742 INFO Starting validation pipeline for block header 807a3a3c at 86.
Nov 14 12:53:34.742 INFO Invalid block header 807a3a3c: Orphan.
Nov 14 12:53:34.743 INFO Starting validation pipeline for block header 1256d4c7 at 87.
Nov 14 12:53:34.743 INFO Invalid block header 1256d4c7: Orphan.
Nov 14 12:53:34.743 INFO Starting validation pipeline for block header 08936e04 at 88.
Nov 14 12:53:34.743 INFO Invalid block header 08936e04: Orphan.
Nov 14 12:53:34.743 INFO Starting validation pipeline for block header 52ce6f55 at 89.
Nov 14 12:53:34.743 INFO Invalid block header 52ce6f55: Orphan.
Nov 14 12:53:34.744 INFO Starting validation pipeline for block header 611a81b3 at 90.
Nov 14 12:53:34.744 INFO Invalid block header 611a81b3: Orphan.
Nov 14 12:53:34.744 INFO Added 0 headers to the header chain.
```

**Comment by ignopeverell** [comment]
It looks like the node is missing block at (n+1) on the other chain, so getting subsequent nodes doesn't help and they're all orphaned. In theory if the other chain is received complete, our chain will switch once more work has been accumulated on the fork. I can take a look at that, seems easy to reproduce.

**Comment by antiochp** [comment]
:+1: let me know if i can help here. I'm digging around getting familiar with the code.

**Comment by ignopeverell** [comment]
Correct me if I'm wrong but I think you may be seeing #226. Or maybe I misunderstood your issue. I've tried variations of the following:

1. Start 2 mining servers s1 and s2.
2. Let them mine a few blocks together, say until height 10.
3. Stop s1 and let s2 mine until 13.
4. Stop s2 and start s1, let it mine until 15.
5. Start s2 again, see what happens.

At step 5 I've had the server re-sync properly every time, as long as the local chain had less work than the other server. Because of #226, I've had to restart the server with less work when one with more work came online tho.

So there may be something more going on with the logs you posted. It'd be good to see how come 82 is already known but 83 is detected as orphaned.

**Comment by antiochp** [comment]
Managed to reproduce this the additional logging in `locate_headers` on the node being sync'd _from_ - 

```
Nov 15 10:33:39.862 DEBG locate_headers: [5eed989c, c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:39.862 DEBG locate_headers: [c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:39.862 DEBG locate_headers: [92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:39.862 DEBG locate_headers: [87a3a171, f6b1e0fc]
Nov 15 10:33:39.862 DEBG locate_headers: [f6b1e0fc]
Nov 15 10:33:39.862 DEBG locate_headers: []
Nov 15 10:33:41.869 DEBG locate_headers: [5eed989c, c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:41.869 DEBG locate_headers: [c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:41.869 DEBG locate_headers: [92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:41.869 DEBG locate_headers: [87a3a171, f6b1e0fc]
Nov 15 10:33:41.870 DEBG locate_headers: [f6b1e0fc]
Nov 15 10:33:41.870 DEBG locate_headers: []
Nov 15 10:33:43.876 DEBG locate_headers: [5eed989c, c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:43.877 DEBG locate_headers: [c93b409c, 92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:43.877 DEBG locate_headers: [92b3c091, 87a3a171, f6b1e0fc]
Nov 15 10:33:43.877 DEBG locate_headers: [87a3a171, f6b1e0fc]
Nov 15 10:33:43.877 DEBG locate_headers: [f6b1e0fc]
Nov 15 10:33:43.877 DEBG locate_headers: []
```

So the node that is sync'ing is asking for several headers but because we are increasing the gap between each header hash (n*2 each time) I think we're ending up asking for a gap large enough to skip everything on the other node.

And it just gets into a loop stuck asking for these (none of these hashes are common across both nodes...)

Proposal: can we push the genesis header hash (height 0) into the locator as a fallback in this case?

**Comment by antiochp** [comment]
Easiest (pathological) way to reproduce this is to start each miner up completely independently and let them both mine divergent chains from the genesis block and then get them to try and sync.
Otherwise the `get_locator` is very sensitive to the current height - sometimes is works fine.

**Comment by ignopeverell** [comment]
Fixed by #271
